packages/react/src/hooks/useAuth.ts

```diff
@@ -0,0 +1,56 @@
+import Auth, { CognitoUser } from '@aws-amplify/auth';
+import { Hub } from '@aws-amplify/core';
+import { useEffect, useState } from 'react';
+
+// Exposes relevant CognitoUser properties
+interface AuthUser extends CognitoUser {
+  username: string;
+  attributes: Record<string, string>;
+}
+
+export const useAuth = () => {
+  const [user, setUser] = useState<AuthUser>();
+  const [error, setError] = useState<Error>();
+  const [isLoading, setLoading] = useState<boolean>(true);
+
+  const handleUser = (user: AuthUser | undefined) => {
+    if (typeof user === 'undefined') {
+      setError(new Error('Current authenticated user is not available'));
+    } else {
+      setUser(user);
+    }
+  };
+
+  const handleAuth = ({ payload }) => {
+    switch (payload.event) {
+      case 'signIn':
+        return handleUser(payload.data);
+      case 'signOut':
+        return handleUser(undefined);
+      default:
+        break;
+    }
+  };
+
+  const fetch = () => {
+    Auth.currentAuthenticatedUser()
+      .then(handleUser)
+      .catch(setError)
```

Comment:
I'm thinking about what happens in federated flows 🤔 

If this fails the first time, then `setError(error)` & `setLoading(false)` is called:
```js
{
  error: someError,
  isLoading: false,
  user: undefined,
}
```


But then if the auth event eventuaally succeeds and `Hub.listen('auth', handleAuth)` is called, then `handleUser(user)` & `setLoading(false)` is called:

```js
{
  error: someError, // Oh no!
  isLoading: false,
  user: someUser,
}
```

This seems to have the following behavior:

- `loading` is `true` on initial load, but not on multiple `handleAuth` invocations. (I assume this is desirable for an app to wait until we resolve/reject)
- `error`s aren't cleared when `setUser` is called.

You may consider using a single state object or reducer for `{ error, isLoading, user }` to solve for interdependent values:

```js
const [result, setResult] = useState<AuthResult>({
  error: undefined,
  isLoading: true,
  user: undefined,
});

...

Auth.currentAuthenticatedUser()
  .then(user => setResult({ isLoading: false, user })
  .catch(error => setResult({ error, isLoading: false })
```

Reply:
really like this approach, will change in next commit :)

Reply:
Updated. Thanks for your suggestions <3